code/game/objects/structures/safe.dm

```diff
@@ -211,96 +212,80 @@ GLOBAL_LIST_EMPTY(safes)
 	if(total_ticks == 1 || prob(10))
 		to_chat(user, "<span class='italics'>You hear a [pick(sounds)] from [src].</span>")
 
-/obj/structure/safe/Topic(href, href_list)
-	if(..())
-		return TRUE
-
-	var/mob/user = usr
-	if(!user.IsAdvancedToolUser() && !isobserver(user))
-		to_chat(user, "<span class='warning'>You're not able to operate the safe.</span>")
+/obj/structure/safe/proc/start_turning_knob(mob/user, canhear, ticks, left)
+	if(spinning || open)
 		return
 
-	var/canhear = FALSE
-	if(ishuman(user))
-		var/mob/living/carbon/human/H = user
-		if(H.can_hear() && H.is_in_hands(/obj/item/clothing/accessory/stethoscope))
-			canhear = TRUE
-
-	if(href_list["open"])
-		if(check_unlocked() || open || broken)
-			to_chat(user, "<span class='notice'>You [open ? "close" : "open"] [src].</span>")
-			open = !open
-			update_icon()
-		else
-			to_chat(user, "<span class='warning'>You can't open [src], as its lock is engaged!</span>")
-
-		. = TRUE
-
-	if(href_list["turnright"])
-		if(open)
-			return
-
-		if(broken)
-			to_chat(user, "<span class='warning'>The dial will not turn, as the mechanism is destroyed.</span>")
-			return
+	if(broken)
+		to_chat(user, "<span class='warning'>The dial will not turn, as the mechanism is destroyed.</span>")
+		return
 
-		var/ticks = text2num(href_list["turnright"])
-		for(var/i = 1 to ticks)
-			dial = WRAP(dial - 1, 0, 100)
+	spinning = TRUE
 
-			var/invalid_turn = current_tumbler_index % 2 == 0 || current_tumbler_index > number_of_tumblers
-			if(invalid_turn) // The moment you turn the wrong way or go too far, the tumblers reset
-				current_tumbler_index = 1
+	var/wrap_delta = left ? 1 : -1
+	for(var/i = 1 to ticks)
+		dial = WRAP(dial + wrap_delta, 0, 100)
 
-			if(!invalid_turn && dial == tumblers[current_tumbler_index])
-				notify_user(user, canhear, list("tink", "krink", "plink"), ticks, i)
-				current_tumbler_index++
-			else
-				notify_user(user, canhear, list("clack", "scrape", "clank"), ticks, i)
+		var/even_tumbler = current_tumbler_index % 2 == 0
+		var/invalid_turn = (left && !even_tumbler) || (!left && even_tumbler) || current_tumbler_index > number_of_tumblers
+		if(invalid_turn) // The moment you turn the wrong way or go too far, the tumblers reset
+			current_tumbler_index = 1
 
-			sleep(1)
-			check_unlocked(user, canhear)
-			SSnanoui.update_uis(src)
+		if(!invalid_turn && dial == tumblers[current_tumbler_index])
+			notify_user(user, canhear, list("tonk", "krunk", "plunk"), ticks, i)
+			current_tumbler_index++
+		else
+			notify_user(user, canhear, list("click", "chink", "clink"), ticks, i)
 
-		. = TRUE
+	addtimer(CALLBACK(src, .proc/finish_turning_knob, user, canhear), 1)
```

Comment:
If I'm reading this right, this only updates the UI once, after all the turning is done, instead of on each tick? And there's little reason to have a callback here, and not just call `finish_turning_knob` directly?

The previous code had it turn gradually(-ish.. as much as sleep(1) can be called gradual). I'm not sure if having `sleep(1)` in here messes with `tgui_act` but if so, you probably need the whole `start_turning_knob` be in a callback, and not just the `finish_turning_knob` part of it.